Source/Randomizer.cpp

```diff
@@ -132,16 +132,50 @@ void Randomizer::Randomize() {
     RandomizeChallenge();
 
     // Content swaps -- must happen before squarePanels
-    Randomize(upDownPanelsSetZero, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetOne, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetTwo, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetThree, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
-    Randomize(leftForwardRightPanelsSetOne, SWAP::LINES | SWAP::COLORS);
-    Randomize(leftForwardRightPanelsSetTwo, SWAP::LINES | SWAP::COLORS);
-
-    Randomize(quarryLaserOptions, SWAP::LINES | SWAP::COLORS);
-    Randomize(squarePanels, SWAP::LINES | SWAP::COLORS);
+    if (_doubleRandomizer) {
+        // The pool that Tutorial Back Left is in has some panels that may not
+        // be solveable in the down position after using Sigma's randomizer. We
+        // do not want to swap any such panel with Tutorial Back Left, because
+        // that would make it impossible to exit Tutorial.
+        Randomize(upDownPanelsSetThreeDoubleMode, SWAP::LINES | SWAP::COLORS);
+        // The four pivot panels in Treehouse must be solveable in the up, left,
+        // and right positions. However, the other panels in the pools those
+        // panels are found in may not be solveable in all three directions
+        // after using Sigma's randomizer. For safety, we will only swap the
+        // pivot panels amongst themselves.
+        Randomize(treehousePivotSet, SWAP::LINES | SWAP::COLORS);
+        // This will additionally shuffle the UTM perspective puzzles amongst
+        // themselves.
+        Randomize(utmPerspectiveSet, SWAP::LINES | SWAP::COLORS);
+        // In order to prevent a situation where access to a Symmetry Laser
+        // Yellow is blocked by its corresponding Blue panel, we will only
+        // shuffle these six panels amongst themselves.
+        Randomize(symmetryLaserYellows, SWAP::LINES | SWAP::COLORS);
+        Randomize(symmetryLaserBlues, SWAP::LINES | SWAP::COLORS);
+        // Prevent the fourth UTM perspective puzzle from being shuffled, as it
+        // is unknown whether any of the other puzzles in these pools can be
+        // solved there after using Sigma's randomizer.
+        Randomize(upDownPanelsSetZeroDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetOneDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetTwoDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
+        // Many puzzles either crash the game or do not solve properly when
+        // swapped with Swamp Entry. To make things simpler, we will just remove
+        // that panel from both pools it is found in.
+        Randomize(quarryLaserOptionsDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(squarePanelsDoubleMode, SWAP::LINES | SWAP::COLORS);
```

Comment:
For simple modifications like this, can we just remove the dangerous panels? That way we don't have to maintain two similar (but long) panel lists.
```suggestion
        // Many puzzles either crash the game or do not solve properly when
        // swapped with Swamp Entry. To make things simpler, we will just remove
        // that panel from both pools it is found in.
        quarryLaserOptions.remove(0x00A52); squarePanels.remove(0x00A52), // Symmetry Island Laser Yellow 1
        quarryLaserOptions.remove(0x00A57); squarePanels.remove(0x00A57), // Symmetry Island Laser Yellow 2
        quarryLaserOptions.remove(0x00A5B); squarePanels.remove(0x00A5B), // Symmetry Island Laser Yellow 3
        quarryLaserOptions.remove(0x00A61); squarePanels.remove(0x00A61), // Symmetry Island Laser Blue 1
        quarryLaserOptions.remove(0x00A64); squarePanels.remove(0x00A64), // Symmetry Island Laser Blue 2
        quarryLaserOptions.remove(0x00A68); squarePanels.remove(0x00A68), // Symmetry Island Laser Blue 3
        quarryLaserOptions.remove(0x0056E); squarePanels.remove(0x0056E), // Swamp Entry
        Randomize(quarryLaserOptionsDoubleMode, SWAP::LINES | SWAP::COLORS);
        Randomize(squarePanelsDoubleMode, SWAP::LINES | SWAP::COLORS);
```

Reply:
The reason I didn't do this was because it'll mess things up if a player uses double randomizer mode, then starts a new game and uses the randomizer again with double randomizer mode unchecked. We could still do something like this though by copying each set into a local variable, removing the dangerous panels, and then using those sets for randomizing. I agree that it would be better than duplicating the huge lists.

Reply:
Changed & tested (for this and for the other comments).